GlobalISel: Use G_UADDE when narrowing G_UMULH

arsenm opened the pull request:
This greatly shrinks the AMDGPU div64 expansion.

Instead of adding a zext of the condition output, add a zero and use
the carry in to G_UADDE. This is closer to how the DAG expansion using
umulh does it, and it seems more natural to leave the boolean output
as a boolean input. We should have a combine to form G_UADDE from this
pattern, but the legalizer shouldn't create extra work for the
combiner if it can help it.

The Mips cases are regressions, but the DAG lowering for muli128 seems
to not use the expansion involving MULHU/MULHS at all. The DAG output
is radically different than GlobalISel as-is, so it seems like Mips
should be using a different legalization strategy here to begin with.

The RISCV legalizer tests look worse for the mul i96 case, but those
didn't exist when I wrote this patch and forgot about it 4 years ago,
so I haven't really looked into why. We've entered the age where most tests
should just be using IR, so I don't  know if this matters or not (the IR mul
test doesn't seem to cover i96)

arsenm committed:
GlobalISel: Use G_UADDE when narrowing G_UMULH

This greatly shrinks the AMDGPU div64 expansion.

Instead of adding a zext of the condition output, add a zero and use
the carry in to G_UADDE. This is closer to how the DAG expansion using
umulh does it, and it seems more natural to leave the boolean output
as a boolean input. We should have a combine to form G_UADDE from this
pattern, but the legalizer shouldn't create extra work for the
combiner if it can help it.

The Mips cases are regressions, but the DAG lowering for muli128 seems
to not use the expansion involving MULHU/MULHS at all. The DAG output
is radically different than GlobalISel as-is, so it seems like Mips
should be using a different legalization strategy here to begin with.

The RISCV legalizer tests look worse for the mul i96 case, but those
didn't exist when I wrote this patch and forgot about it 4 years ago,
so I haven't really looked into why. We've entered the age where most tests
should just be using IR, so I don't  know if this matters or not (the IR mul
test doesn't seem to cover i96)

arsenm commented:
* **#97194** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/97194?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @arsenm and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

tschuett commented:
Did you mean `the carry in to G_UADDO`?

For the G_**DDOs, there are combines relying on known bits, but the G**DDEs are really bad to combine.

tschuett commented:
https://discourse.llvm.org/t/pattern-matching-g-su-add-oe-and-g-su-sub-oe/79208

tschuett commented:
Why is AArch64 not affected? Missing test coverage?

tschuett commented:
For G_UADDE, we could try to combine it down to add + constant 1: https://github.com/llvm/llvm-project/blob/66d86a63e7f10905fd3e6104cebceeb53851ff1d/llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp#L7293

```
(sum, carryout) = G_UADDE (x, y, carryin)
```
If x + y always overflows, ...

tschuett commented on llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:
```diff
@@ -5716,6 +5716,7 @@ void LegalizerHelper::multiplyRegisters(SmallVectorImpl<Register> &DstRegs,
                                         ArrayRef<Register> Src1Regs,
                                         ArrayRef<Register> Src2Regs,
                                         LLT NarrowTy) {
+  const LLT S1 = LLT::scalar(1);
```
Why `const` ?

tschuett commented on llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:
```diff
@@ -5752,15 +5755,15 @@ void LegalizerHelper::multiplyRegisters(SmallVectorImpl<Register> &DstRegs,
+        auto Uadde =
+            B.buildUAdde(NarrowTy, S1, FactorSum, Factors[i], Uaddo.getReg(1));
+        FactorSum = Uadde.getReg(0);
+        CarrySum = B.buildUAdde(NarrowTy, S1, CarrySum, Zero, Uadde.getReg(1))
```
What is the semantic difference to an `G_UADDO`? One parameter of the G_UADDE is zero. Ok, maybe zext the carryin.

arsenm replied:
It's mainly a structural difference. G_UADDE has the carry-in boolean flag, G_UADDO does not

arsenm commented:
> "We should have a combine to form G_UADDE from this pattern,". This will be answered in the Combiner.rst file. Are we allowed to write abstraction raising combines?

What do you mean by `abstraction raising` here? 

> In this case, I am strongly opposed. I don't want the combiner to create more legalization artefacts.

I don't follow. No legalization artifacts are mentioned here. It's an instruction, typically used for legalization, which should have a rich set of optimization combines on it. It is not a legalization artifact, nor are legalization artifacts something to be avoided

arsenm commented:
> > In this case, I am strongly opposed. I don't want the combiner to create more legalization artefacts.
> 
> I don't follow. No legalization artifacts are mentioned here. It's an instruction, typically used for legalization, which should have a rich set of optimization combines on it. It is not a legalization artifact, nor are legalization artifacts something to be avoided

Plus this is a separate TODO combine, we're avoiding / delaying doing it by introducing UADDE from the beginning

arsenm commented:
> Why is AArch64 not affected? Missing test coverage?

My guess is since it has a 64-bit multiply, this will typically be unused. The test coverage for really wide multiplies is probably lacking

tschuett commented:
https://github.com/llvm/llvm-project/issues/92309

abstraction raising:  combine primitive ops into a higher abstraction:
https://github.com/llvm/llvm-project/blob/66d86a63e7f10905fd3e6104cebceeb53851ff1d/llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp#L4523

The same would be combining from more primitives ops into G_UADDE.

arsenm commented:
> The same would be combining from more primitives ops into G_UADDE.

This is an argument against having any optimizations. Most optimizations are of the form take-simpler-instructions and form more-complex instructions.

tschuett commented:
1. Feel free to extend https://github.com/llvm/llvm-project/issues/92309
2. We combine select(cmp) into min/max.
3. We combine select into binary operators.

arsenm commented:
> 1. Feel free to extend [Document GlobalISel's combiner guidelines #92309](https://github.com/llvm/llvm-project/issues/92309)
> 2. We combine select(cmp) into min/max.
> 3. We combine select into binary operators.

These are all target dependent optimization choices. Ideally we would have a library of combines in both directions, and the target is free to enable whichever set it wants. We could have an IR-esque canonicalization and de-canonicalization sequence of combiner passes

arsenm commented:
> > Why is AArch64 not affected? Missing test coverage?
> 
> My guess is since it has a 64-bit multiply, this will typically be unused. The test coverage for really wide multiplies is probably lacking

Yes, missing tests. It has 64-bit umulh, so I have to go all the way to an i160 multiply before I see this expansion kick in

arsenm committed:
AArch64/GlobalISel: Add IR test for scalar multiplies

I didn't see an existing, simple test covering mul. There
appeared to be no wide multiply tests.

arsenm commented:
> Did you mean `the carry in to G_UADDO`?

No. Previously this was doing add 0/1 for overflow to G_UADDO, and now it uses the carry-in input of G_UADDE

jayfoad commented on llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:
```diff
@@ -5755,15 +5758,15 @@ void LegalizerHelper::multiplyRegisters(SmallVectorImpl<Register> &DstRegs,
-        FactorSum = Uaddo.getReg(0);
-        MachineInstrBuilder Carry = B.buildZExt(NarrowTy, Uaddo.getReg(1));
-        CarrySum = B.buildAdd(NarrowTy, CarrySum, Carry).getReg(0);
+        auto Uadde =
+            B.buildUAdde(NarrowTy, S1, FactorSum, Factors[i], Uaddo.getReg(1));
```
No, this one should stay as a uaddo. The carry output from the original uaddo (outside the loop) should go into CarrySum once, but you're adding it into FactorSum every time around this loop.

tschuett commented:
> > > Why is AArch64 not affected? Missing test coverage?
> > 
> > 
> > My guess is since it has a 64-bit multiply, this will typically be unused. The test coverage for really wide multiplies is probably lacking
> 
> Yes, missing tests. It has 64-bit umulh, so I have to go all the way to an i160 multiply before I see this expansion kick in

Thanks!
